### ISSUE TITLE
xtext 2.17: Tons of exceptions logged at ProjectOpenedOrClosedListener.processClosedProjects

### ISSUE THREAD

**Issue by trancexpress** [eclipse/xtext-eclipse#1066]
We are trying out `xtext 2.17.1` as an update from `xtext 2.16` for our product.

During our ARTs we see tons and tons of exceptions with the following stack trace:

```
Unexpected log from plugin: org.eclipse.core.runtime, message: 'An internal error occurred during: "Removing project .org.eclipse.jdt.core.external.folders from Xtext index.".'
org.eclipse.core.runtime.AssertionFailedException: null argument:
	at org.eclipse.core.runtime.Assert.isNotNull(Assert.java:88)
	at org.eclipse.core.runtime.Assert.isNotNull(Assert.java:76)
	at org.eclipse.core.runtime.Path.initialize(Path.java:644)
	at org.eclipse.core.runtime.Path.<init>(Path.java:211)
	at org.eclipse.core.internal.resources.WorkspaceRoot.getProject(WorkspaceRoot.java:148)
	at abc.XYZ.doUpdate(XYZ.java:300)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.update(AbstractBuilderState.java:116)
	at org.eclipse.xtext.builder.impl.ProjectOpenedOrClosedListener.processClosedProjects(ProjectOpenedOrClosedListener.java:275)
	at org.eclipse.xtext.builder.impl.ProjectOpenedOrClosedListener$RemoveProjectsJob.runInWorkspace(ProjectOpenedOrClosedListener.java:104)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:42)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

After investigating, we observed that we have a class which overrides the method  `org.eclipse.xtext.builder.clustering.ClusteringBuilderState.doUpdate(BuildData, ResourceDescriptionsData, IProgressMonitor)`; the first call of our implementation is:

```
IProject project = workspace.getRoot().getProject(buildData.getProjectName());
```

This code is called `null` with project name by `org.eclipse.xtext.builder.impl.ProjectOpenedOrClosedListener.processClosedProjects(IProgressMonitor)` in the following code block:

```
try {
	ResourceSet resourceSet = createResourceSet();
	BuildData buildData = new BuildData(null, resourceSet, task.getToBeBuilt(), queuedBuildData, false,
			BuildManagerAccess::requestBuild);
	getBuilderState().update(buildData, progress.newChild(1));
} catch (Error | RuntimeException e) {
	task.reschedule();
	throw e;
} finally {
	monitor.done();
}
```

We have 2 issues with this:

1. Looks like the code in `ProjectOpenedOrClosedListener` is meant to bundle up several projects and process them in one go. The delta should contain the names of the projects involved instead of `null`.
2. `NullPointerException` is a `RuntimeException` and is a fairly common exception. Rescheduling the job on a `NullPointerException` is probably not the best idea, since an implementer which threw a NPE will likely do the same on the next call. Our test times have bloated to 3 times of the usual test times, highly likely due to this rescheduling alone.

**Comment by cdietrich** [comment]
cc @szarnekow

**Comment by szarnekow** [comment]
I'm on it.

**Comment by cdietrich** [comment]
fixed in 2.18
